src/service/handlers/prepared.rs

```diff
@@ -276,19 +284,73 @@ pub(crate) async fn do_put_prepared_statement_update(
         ));
     }
 
-    let parameter_sets = SwanFlightSqlService::collect_parameter_sets(request).await?;
+    // Collect RecordBatches first - we can only consume the stream once
+    let batches = SwanFlightSqlService::collect_record_batches(request).await?;
+    
+    // No batches provided - execute with empty parameters
+    if batches.is_empty() {
+        info!(
+            handle = %handle,
+            sql = %sql,
+            "executing prepared statement update without parameters"
+        );
 
-    let parameter_set_count = parameter_sets.len();
+        let session_clone = Arc::clone(&session);
+        let affected_rows = tokio::task::spawn_blocking(move || {
+            SwanFlightSqlService::execute_statement_batches(&sql, &[], &session_clone)
+        })
+        .await
+        .map_err(SwanFlightSqlService::status_from_join)?
+        .map_err(SwanFlightSqlService::status_from_error)?;
 
-    info!(
-        handle = %handle,
-        sql = %sql,
-        parameter_sets = parameter_set_count,
-        "executing prepared statement update"
-    );
+        info!(
+            handle = %handle,
+            affected_rows,
+            "prepared statement update complete"
+        );
+        return Ok(affected_rows);
+    }
 
+    // Try to use appender optimization via session layer
     let session_clone = Arc::clone(&session);
+    let sql_clone = sql.clone();
+    let batches_clone = batches.clone();
+    
+    let affected_rows = tokio::task::spawn_blocking(move || {
+        session_clone
+            .execute_with_record_batches(&sql_clone, batches_clone)
+            .map(|rows| rows as i64)
+    })
+    .await
+    .map_err(SwanFlightSqlService::status_from_join)?;
+    
+    // If appender optimization succeeded, return the result
+    if let Ok(rows) = affected_rows {
+        info!(
+            handle = %handle,
+            affected_rows = rows,
+            "statement executed with RecordBatch optimization"
+        );
+        return Ok(rows);
+    }
+
+    // Fallback: convert RecordBatches to parameters for non-INSERT statements
```

Comment:
in which case it will fail? we can't just fallback everything!

if you don't have a very solid reason, delete all fallback bullshit.